### PR TITLE
chore(workspace): validate filename that are illegal in SQLite

### PR DESCRIPTION
**Pull request by hikchoi**
# chore(workspace): validate filename that are illegal in SQLite

This PR:
- Extends the file name validation logic to capture illegal SQLite characters.
- Enhance warning / error messages to be more specific about what is invalid.
- Add doctor action `fixInvalidFileNames` on cli and extension
- 

# Pull Request Checklist
### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [x] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [x] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [x] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [~] display is correct for following dimensions
    - [~] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [~] lg: screen ≥ 992px
    - [~] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [~] display is correct for following browsers (across the various dimensions)
    - [~] safari
    - [~] firefox
    - [~] chrome

## Docs
- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by hikchoi** [comment]
docs: https://github.com/dendronhq/dendron-site/pull/639

File: packages/engine-server/src/doctor/service.ts [hunk 1]
```diff
@@ -593,6 +597,59 @@ export class DoctorService implements Disposable {
...
+          notes,
+          noteDicts: {
+            notesById,
+            notesByFname,
+          },
+        });
+        resp = stats;
+
+        if (canRename.length > 0 && !opts.quiet) {
+          console.log("Found invalid filename in notes:\n");
```
Comment (kevinslin) [inline packages/engine-server/src/doctor/service.ts:611]: ideally, we wouldn't be using `console.log` here but the [print](https://github.com/dendronhq/dendron/blob/master/packages/dendron-cli/src/commands/base.ts#L255:L255) function from the `CLICommand`. can we pass it in as an optional parameter in the constructor? by default, its a no-op.

File: packages/engine-server/src/doctor/service.ts [hunk 2]
```diff
@@ -593,6 +597,59 @@ export class DoctorService implements Disposable {
...
+          canRename.forEach((item) => {
+            const { note, resp, cleanedFname } = item;
+            const { fname, vault } = note;
+            const vaultName = VaultUtils.getName(vault);
+            console.log(
+              `Note "${fname}" in ${vaultName} (reason: ${resp.reason})`
+            );
+            console.log(`  Can be automatically fixed to "${cleanedFname}"`);
```
Comment (kevinslin) [inline packages/engine-server/src/doctor/service.ts:619]: extra space

File: packages/engine-server/src/doctor/service.ts (hunk 1 above)
Comment (kevinslin) [reply packages/engine-server/src/doctor/service.ts:611]: also apply to other `console.log` statements

File: packages/plugin-core/src/commands/Doctor.ts [hunk 3]
```diff
@@ -334,10 +414,11 @@ export class DoctorCommand extends BasicCommand<CommandOpts, CommandOutput> {
     return engine;
   }
 
-  addAnalyticsPayload(opts: CommandOpts) {
+  addAnalyticsPayload(opts: CommandOpts, out: CommandOutput) {
     return {
       action: opts.action,
       scope: opts.scope,
+      ...out.extra,
```
Comment (kevinslin) [inline packages/plugin-core/src/commands/Doctor.ts:421]: we should pick what properties we are adding but not extract everything from an `any` parameter

**Review by kevinslin — CHANGES_REQUESTED** [review]

**Review by jonathanyeung — APPROVED** [review]
lgtm apart from existing comments.

File: packages/engine-server/src/doctor/service.ts (hunk 2 above)
Comment (hikchoi) [reply packages/engine-server/src/doctor/service.ts:619]: it's an intentional indentation

**Review by kevinslin — APPROVED** [review]

**Pull request merged**